Create typings.d.ts

edorivai opened the pull request:
This is the TypeScript typing that we're using internally. Please let me know if changes are required.

swese44 commented on typings.d.ts:
```diff
@@ -0,0 +1,9 @@
+declare module 'react-media' {
+  interface MediaProps {
+    query: string;
```
`query` should be a string or an object with string keys and values of number, string or boolean. looks like react-media will pass a string query to `window.matchMedia(query)`, and will pass an object query directly to json2mq:
https://github.com/akiran/json2mq

maybe this:
```
export interface MediaQueryObject {
    [id: string]: boolean | number | string;
}

export interface MediaProps {
    query: string | MediaQueryObject;
    ...
}
```

swese44 replied:
looks like `query` can also accept an array of objects, so maybe this is better:

```
export interface MediaQueryObject {
    [id: string]: boolean | number | string;
}

export interface MediaProps {
    query: string | MediaQueryObject | MediaQueryObject[];
    ...
}
```

swese44 commented on typings.d.ts:
```diff
@@ -0,0 +1,9 @@
+  interface MediaProps {
+    query: string;
+    defaultMatches?: boolean;
+    children: ((matches: boolean) => JSX.Element) | JSX.Element
```
`children` should be optional, and i believe these should specify React.ReactNode instead of JSX.Element

swese44 commented on typings.d.ts:
```diff
@@ -0,0 +1,9 @@
+    query: string;
+    defaultMatches?: boolean;
+    children: ((matches: boolean) => JSX.Element) | JSX.Element
+  }
```
we're also missing the optional `render` prop:
`render?: () => React.ReactNode;`

swese44 commented on typings.d.ts:
```diff
@@ -0,0 +1,9 @@
+declare module 'react-media' {
+  interface MediaProps {
```
don't forget to export the props interface so other components can access it

edorivai commented:
@swese44 Thanks for all the hints! I think I've fixed all of the issues.

mjackson commented:
If you'd like to include this in the npm package you'll need to add it to the `files` array in `package.json`.

edorivai commented:
I've renamed the TS file to `index.d.ts`, and added it to the `files` array. Hope everything is set up properly!

One final question: the typings now depend on `@types/react` being installed as well, as it uses the global `React.ReactNode` and `React.Component` types. I was wondering if anyone knows the best practices around dependencies between TS typing files. I don't really have any experience with this.

unindented commented on index.d.ts:
```diff
@@ -0,0 +1,14 @@
+declare module 'react-media' {
```
This line is unnecessary when you bundle typings with the package like we're doing here. The module name will be the one in `package.json`.

unindented commented on package.json:
```diff
@@ -8,7 +8,8 @@
     "esm",
-    "umd"
+		"umd",
+		"index.d.ts"
```
You seem to be using tabs, but the rest of the file doesn't.

edorivai replied:
Oops, my bad

unindented commented:
@edorivai left two comments.

Regarding dependencies, you should import them in your `index.d.ts` file:

```ts
import { Component, ReactNode } from 'react';
```

edorivai commented:
@unindented Fixed both issues, and converted index.d.ts to spaces as well :nerd_face:

unindented commented:
The line `declare module 'react-media'` is unnecessary in this instance, cause you're bundling the types with the module itself.

edorivai commented:
Oh really, didn't know that. Nice!

edorivai commented:
Btw, once everything looks okay, I'll happily squash everything into 1 commit, if you guys prefer that.

farzadmf commented:
Hi,

Is this going to be merged? Is there a workaround to use the library in Typescript?

Thanks

edorivai commented:
@farzadmf workaround would be to copy/paste this file into your project:
https://github.com/edorivai/react-media/blob/37aa81c350eac95bc2d8779400959abbdf87767b/index.d.ts

farzadmf commented:
Thank you @edorivai , I thought of that 👍 ; just wanted to check if there's a "more official" way to do it :)

unindented commented:
@mjackson any chance of getting this merged, and a new release published?
